`./pants --loop ...` with failing command continuously reruns, forever

**Describe the bug**

When running `./pants --loop ...` with a command that has an error, pants will run and rerun the command forever. Examples of failing commands:

- `./pants --loop fmt ::` with black configured when there's a Python source file with invalid syntax (spelled out below)
- any pants command that needs to run a `shell_command`/`adhoc_tool` with exit code != 0 https://github.com/pantsbuild/pants/issues/18275
- running any goal on a target that doesn't exist https://github.com/pantsbuild/pants/issues/18276
- running any goal when a BUILD file is invalid https://github.com/pantsbuild/pants/issues/18277


Reproducer of the `black` example (see linked issues for other simpler reproducers): https://gist.github.com/huonw/55bf0340a467881dc8511346f5ad2d5a has a single bad file `broken.py`. Running:

```shell
git clone git@gist.github.com:55bf0340a467881dc8511346f5ad2d5a.git
cd 55bf0340a467881dc8511346f5ad2d5a
./pants --loop fmt ::
```

This gives endless copies of the black error message:

```
09:17:34.49 [ERROR] 1 Exception encountered:

  ProcessExecutionFailure: Process 'Run Black on 1 file.' failed with exit code 123.
stdout:

stderr:
error: cannot format broken.py: Cannot parse: 1:8: invalid formatting

Oh no! 💥 💔 💥
1 file failed to reformat.



Use `--keep-sandboxes=on_failure` to preserve the process chroot for inspection.
```
```
09:17:34.59 [ERROR] 1 Exception encountered:

  ProcessExecutionFailure: Process 'Run Black on 1 file.' failed with exit code 123.
stdout:

stderr:
error: cannot format broken.py: Cannot parse: 1:8: invalid formatting

Oh no! 💥 💔 💥
1 file failed to reformat.



Use `--keep-sandboxes=on_failure` to preserve the process chroot for inspection.
```
etc.

**Pants version**
- 2.12.0
- 2.14.0a0
- 2.16.0.dev6

**OS**
macOS

**Additional info**
N/A

--- a/src/rust/engine/graph/src/entry.rs
+++ b/src/rust/engine/graph/src/entry.rs
@@ -156,7 +156,7 @@ impl<N: Node> AsRef<N::Item> for EntryResult<N> {
   }
 }
 
-pub type NodeResult<N> = Result<(<N as Node>::Item, Generation), <N as Node>::Error>;
+pub type NodeResult<N> = (Result<<N as Node>::Item, <N as Node>::Error>, Generation);
 
 #[derive(Debug)]
 pub enum EntryState<N: Node> {
@@ -170,6 +170,7 @@ pub enum EntryState<N: Node> {
   NotStarted {
     run_token: RunToken,
     generation: Generation,
+    pollers: Vec<oneshot::Sender<()>>,
     previous_result: Option<EntryResult<N>>,
   },
   // A node that is running. A running node that has been marked dirty re-runs rather than
@@ -206,6 +207,7 @@ impl<N: Node> EntryState<N> {
     EntryState::NotStarted {
       run_token: RunToken::initial(),
       generation: Generation::initial(),
+      pollers: Vec::new(),
       previous_result: None,
     }
   }
@@ -259,27 +261,38 @@ impl<N: Node> Entry<N> {
   pub async fn poll(&self, context: &N::Context, last_seen_generation: Generation) {
     let recv = {
       let mut state = self.state.lock();
-      match *state {
+      let pollers = match *state {
         EntryState::Completed {
           ref result,
           generation,
           ref mut pollers,
           ..
         } if generation == last_seen_generation && result.poll_should_wait(context) => {
-          // The Node is currently clean with the observed generation: add a poller on the
-          // Completed node that will be notified when it is dirtied or dropped. If the Node moves
-          // to another state, the received will be notified that the sender was dropped, and it
-          // will be converted into a successful result.
-          let (send, recv) = oneshot::channel();
-          pollers.push(send);
-          recv
+          // The Node is clean in this context, and the last seen generation matches.
+          pollers
+        }
+        EntryState::NotStarted {
+          generation,
+          ref mut pollers,
+          ..
+        } if generation == last_seen_generation => {
+          // The Node has not yet been started, but the last seen generation matches. This
+          // means that an error occurred on a previous run of the node, but it has already been
+          // observed by the caller.
+          pollers
         }
         _ => {
           // The generation didn't match or the Node wasn't Completed. It should be requested
           // without waiting.
           return;
         }
-      }
+      };
+
+      // Add a poller on the node that will be notified when it is dirtied or dropped. If the Node
+      // moves to another state, the receiver will be notified that the sender was dropped.
+      let (send, recv) = oneshot::channel();
+      pollers.push(send);
+      recv
     };
     // Wait outside of the lock.
     let _ = recv.await;
@@ -308,7 +321,7 @@ impl<N: Node> Entry<N> {
     generation: Generation,
     previous_dep_generations: Option<Vec<Generation>>,
     previous_result: Option<EntryResult<N>>,
-  ) -> (EntryState<N>, AsyncValueReceiver<NodeResult<N>>) {
+  ) -> (EntryState<N>, AsyncValueReceiver<NodeResult<N>>, Generation) {
     // Increment the RunToken to uniquely identify this work.
     let run_token = run_token.next();
     let context = context_factory.clone_for(entry_id);
@@ -359,7 +372,7 @@ impl<N: Node> Entry<N> {
         abort_item = sender.aborted() => {
           if let Some(res) = abort_item {
             // We were aborted via terminate: complete with the given res.
-            Some(res.map(|v| v.0))
+            Some(res.0)
           } else {
             // We were aborted via drop: exit.
             context2
@@ -387,6 +400,7 @@ impl<N: Node> Entry<N> {
         is_cleaning,
       },
       receiver,
+      generation,
     )
   }
 
@@ -409,10 +423,18 @@ impl<N: Node> Entry<N> {
     // cases we return early without swapping the state of the Node.
     match *state {
       EntryState::Running {
-        ref pending_value, ..
+        ref pending_value,
+        generation,
+        ..
       } => {
         if let Some(receiver) = pending_value.receiver() {
-          return async move { receiver.recv().await.ok_or_else(N::Error::invalidated)? }.boxed();
+          return async move {
+            receiver
+              .recv()
+              .await
+              .unwrap_or_else(|| (Err(N::Error::invalidated()), generation.next()))
+          }
+          .boxed();
         }
         // Else: this node was just canceled: fall through to restart it.
       }
@@ -421,17 +443,19 @@ impl<N: Node> Entry<N> {
         generation,
         ..
       } if result.is_clean(context) => {
-        return future::ready(Ok((result.as_ref().clone(), generation))).boxed();
+        return future::ready((Ok(result.as_ref().clone()), generation)).boxed();
       }
       _ => (),
     };
 
     // Otherwise, we'll need to swap the state of the Node, so take it by value.
-    let (next_state, receiver) = match mem::replace(&mut *state, EntryState::initial()) {
+    let (next_state, receiver, generation) = match mem::replace(&mut *state, EntryState::initial())
+    {
       EntryState::NotStarted {
         run_token,
         generation,
         previous_result,
+        ..
       }
       | EntryState::Running {
         run_token,
@@ -490,7 +514,13 @@ impl<N: Node> Entry<N> {
     // Swap in the new state, and return the receiver.
     *state = next_state;
 
-    async move { receiver.recv().await.ok_or_else(N::Error::invalidated)? }.boxed()
+    async move {
+      receiver
+        .recv()
+        .await
+        .unwrap_or_else(|| (Err(N::Error::invalidated()), generation.next()))
+    }
+    .boxed()
   }
 
   ///
@@ -521,6 +551,7 @@ impl<N: Node> Entry<N> {
         EntryState::NotStarted {
           run_token: run_token.next(),
           generation,
+          pollers: Vec::new(),
           previous_result,
         }
       }
@@ -582,10 +613,12 @@ impl<N: Node> Entry<N> {
             if let Some(previous_result) = previous_result.as_mut() {
               previous_result.dirty();
             }
-            sender.send(Err(e));
+            generation = generation.next();
+            sender.send((Err(e), generation));
             EntryState::NotStarted {
               run_token: run_token.next(),
               generation,
+              pollers: Vec::new(),
               previous_result,
             }
           }
@@ -597,7 +630,7 @@ impl<N: Node> Entry<N> {
               // Node was re-executed (ie not cleaned) and had a different result value.
               generation = generation.next()
             };
-            sender.send(Ok((next_result.as_ref().clone(), generation)));
+            sender.send((Ok(next_result.as_ref().clone()), generation));
             EntryState::Completed {
               result: next_result,
               pollers: Vec::new(),
@@ -616,7 +649,7 @@ impl<N: Node> Entry<N> {
               self.cacheable_with_output(Some(result.as_ref())),
               has_uncacheable_deps,
             );
-            sender.send(Ok((result.as_ref().clone(), generation)));
+            sender.send((Ok(result.as_ref().clone()), generation));
             EntryState::Completed {
               result,
               pollers: Vec::new(),
@@ -710,6 +743,7 @@ impl<N: Node> Entry<N> {
     *state = EntryState::NotStarted {
       run_token: run_token.next(),
       generation,
+      pollers: Vec::new(),
       previous_result,
     };
   }
@@ -729,14 +763,18 @@ impl<N: Node> Entry<N> {
         ref mut pollers,
         ..
       } => {
-        // Notify all pollers (ignoring any that have gone away.)
-        for poller in pollers.drain(..) {
-          let _ = poller.send(());
-        }
+        // Drop the pollers, which will notify them of a change.
+        pollers.clear();
         result.dirty();
         return;
       }
-      &mut EntryState::NotStarted { .. } => return,
+      &mut EntryState::NotStarted {
+        ref mut pollers, ..
+      } => {
+        // Drop the pollers, which will notify them of a change.
+        pollers.clear();
+        return;
+      }
       &mut EntryState::Running { .. } if !self.node.cacheable() => {
         // An uncacheable node cannot be interrupted.
         return;
@@ -760,6 +798,7 @@ impl<N: Node> Entry<N> {
         EntryState::NotStarted {
           run_token,
           generation,
+          pollers: Vec::new(),
           previous_result,
         }
       }
@@ -776,8 +815,13 @@ impl<N: Node> Entry<N> {
   pub(crate) fn terminate(&mut self, err: N::Error) {
     let state = &mut *self.state.lock();
     test_trace_log!("Terminating node {:?} with {:?}", self.node, err);
-    if let EntryState::Running { pending_value, .. } = state {
-      let _ = pending_value.try_abort(Err(err));
+    if let EntryState::Running {
+      pending_value,
+      generation,
+      ..
+    } = state
+    {
+      let _ = pending_value.try_abort((Err(err), generation.next()));
     };
   }
 

--- a/src/rust/engine/graph/src/lib.rs
+++ b/src/rust/engine/graph/src/lib.rs
@@ -440,7 +440,7 @@ impl<N: Node> Graph<N> {
     src_id: Option<EntryId>,
     context: &N::Context,
     dst_node: N,
-  ) -> Result<(N::Item, Generation), N::Error> {
+  ) -> (Result<N::Item, N::Error>, Generation) {
     // Compute information about the dst under the Graph lock, and then release it.
     let (dst_retry, mut entry, entry_id) = {
       // Get or create the destination, and then insert the dep and return its state.
@@ -478,8 +478,7 @@ impl<N: Node> Graph<N> {
       let context = context.clone();
       loop {
         match entry.get_node_result(&context, entry_id).await {
-          Ok(r) => break Ok(r),
-          Err(err) if err == N::Error::invalidated() => {
+          (Err(err), _) if err == N::Error::invalidated() => {
             let node = {
               let inner = self.inner.lock();
               inner.unsafe_entry_for_id(entry_id).node().clone()
@@ -491,7 +490,7 @@ impl<N: Node> Graph<N> {
             sleep(self.invalidation_delay).await;
             continue;
           }
-          Err(other_err) => break Err(other_err),
+          res => break res,
         }
       }
     } else {
@@ -515,8 +514,8 @@ impl<N: Node> Graph<N> {
     context: &N::Context,
     dst_node: N,
   ) -> Result<N::Item, N::Error> {
-    let (res, _generation) = self.get_inner(src_id, context, dst_node).await?;
-    Ok(res)
+    let (res, _generation) = self.get_inner(src_id, context, dst_node).await;
+    res
   }
 
   ///
@@ -536,7 +535,7 @@ impl<N: Node> Graph<N> {
     token: Option<LastObserved>,
     delay: Option<Duration>,
     context: &N::Context,
-  ) -> Result<(N::Item, LastObserved), N::Error> {
+  ) -> (Result<N::Item, N::Error>, LastObserved) {
     // If the node is currently clean at the given token, Entry::poll will delay until it has
     // changed in some way.
     if let Some(LastObserved(generation)) = token {
@@ -552,8 +551,8 @@ impl<N: Node> Graph<N> {
     };
 
     // Re-request the Node.
-    let (res, generation) = self.get_inner(None, context, node).await?;
-    Ok((res, LastObserved(generation)))
+    let (res, generation) = self.get_inner(None, context, node).await;
+    (res, LastObserved(generation))
   }
 
   ///
@@ -595,10 +594,7 @@ impl<N: Node> Graph<N> {
             .unwrap_or_else(|| panic!("Dependency not present in Graph."))
             .clone();
           async move {
-            let (_, generation) = dep_entry
-              .get_node_result(context, dep_id)
-              .await
-              .map_err(|_| ())?;
+            let (_, generation) = dep_entry.get_node_result(context, dep_id).await;
             if generation == previous_dep_generation {
               // Matched.
               Ok(())

--- a/src/rust/engine/graph/src/tests.rs
+++ b/src/rust/engine/graph/src/tests.rs
@@ -251,23 +251,18 @@ async fn invalidate_randomly() {
 
 #[tokio::test]
 async fn poll_cacheable() {
+  let _logger = env_logger::try_init();
   let graph = empty_graph();
   let context = TContext::new(graph.clone());
 
   // Poll with an empty graph should succeed.
-  let (result, token1) = graph
-    .poll(TNode::new(2), None, None, &context)
-    .await
-    .unwrap();
-  assert_eq!(result, vec![T(0, 0), T(1, 0), T(2, 0)]);
+  let (result, token1) = graph.poll(TNode::new(2), None, None, &context).await;
+  assert_eq!(result.unwrap(), vec![T(0, 0), T(1, 0), T(2, 0)]);
 
   // Re-polling on a non-empty graph but with no LastObserved token should return immediately with
   // the same value, and the same token.
-  let (result, token2) = graph
-    .poll(TNode::new(2), None, None, &context)
-    .await
-    .unwrap();
-  assert_eq!(result, vec![T(0, 0), T(1, 0), T(2, 0)]);
+  let (result, token2) = graph.poll(TNode::new(2), None, None, &context).await;
+  assert_eq!(result.unwrap(), vec![T(0, 0), T(1, 0), T(2, 0)]);
   assert_eq!(token1, token2);
 
   // But polling with the previous token should wait, since nothing has changed.
@@ -279,9 +274,10 @@ async fn poll_cacheable() {
 
   // Invalidating something and re-polling should re-compute.
   graph.invalidate_from_roots(true, |n| n.id == 0);
-  let (result, _) = graph
+  let result = graph
     .poll(TNode::new(2), Some(token2), None, &context)
     .await
+    .0
     .unwrap();
   assert_eq!(result, vec![T(0, 0), T(1, 0), T(2, 0)]);
 }
@@ -298,11 +294,8 @@ async fn poll_uncacheable() {
   };
 
   // Poll with an empty graph should succeed.
-  let (result, token1) = graph
-    .poll(TNode::new(2), None, None, &context)
-    .await
-    .unwrap();
-  assert_eq!(result, vec![T(0, 0), T(1, 0), T(2, 0)]);
+  let (result, token1) = graph.poll(TNode::new(2), None, None, &context).await;
+  assert_eq!(result.unwrap(), vec![T(0, 0), T(1, 0), T(2, 0)]);
 
   // Polling with the previous token (in the same session) should wait, since nothing has changed.
   let request = graph.poll(TNode::new(2), Some(token1), None, &context);
@@ -313,11 +306,30 @@ async fn poll_uncacheable() {
 
   // Invalidating something and re-polling should re-compute.
   graph.invalidate_from_roots(true, |n| n.id == 0);
-  let (result, _) = graph
+  let result = graph
     .poll(TNode::new(2), Some(token1), None, &context)
     .await
+    .0
     .unwrap();
   assert_eq!(result, vec![T(0, 0), T(1, 0), T(2, 0)]);
+}
+
+#[tokio::test]
+async fn poll_errored() {
+  let _logger = env_logger::try_init();
+  let graph = empty_graph();
+  let context = TContext::new(graph.clone()).with_errors(vec![TNode::new(0)].into_iter().collect());
+
+  // Poll should error.
+  let (result, token1) = graph.poll(TNode::new(2), None, None, &context).await;
+  assert_eq!(result.err().unwrap(), TError::Error);
+
+  // Polling with the previous token should wait, since nothing has changed.
+  let request = graph.poll(TNode::new(2), Some(token1), None, &context);
+  match timeout(Duration::from_millis(1000), request).await {
+    Err(Elapsed { .. }) => (),
+    e => panic!("Should have timed out, instead got: {e:?}"),
+  }
 }
 
 #[tokio::test]
@@ -755,6 +767,9 @@ impl Node for TNode {
   async fn run(self, context: TContext) -> Result<Vec<T>, TError> {
     let mut abort_guard = context.abort_guard(self.clone());
     context.ran(self.clone());
+    if context.errors.contains(&self) {
+      return Err(TError::Error);
+    }
     let token = T(self.id, context.salt());
     context.maybe_delay(&self).await;
     let res = match context.dependencies_of(&self) {
@@ -857,6 +872,8 @@ struct TContext {
   // the next smallest node.
   edges: Arc<HashMap<TNode, Vec<TNode>>>,
   delays: Arc<HashMap<TNode, Duration>>,
+  // Nodes which should error when they run.
+  errors: Arc<HashSet<TNode>>,
   non_restartable: Arc<HashSet<TNode>>,
   uncacheable: Arc<HashSet<TNode>>,
   graph: Arc<Graph<TNode>>,
@@ -879,6 +896,7 @@ impl NodeContext for TContext {
       salt: self.salt,
       edges: self.edges.clone(),
       delays: self.delays.clone(),
+      errors: self.errors.clone(),
       non_restartable: self.non_restartable.clone(),
       uncacheable: self.uncacheable.clone(),
       graph: self.graph.clone(),
@@ -913,6 +931,7 @@ impl TContext {
       salt: 0,
       edges: Arc::default(),
       delays: Arc::default(),
+      errors: Arc::default(),
       non_restartable: Arc::default(),
       uncacheable: Arc::default(),
       graph,
@@ -930,6 +949,11 @@ impl TContext {
 
   fn with_delays(mut self, delays: HashMap<TNode, Duration>) -> TContext {
     self.delays = Arc::new(delays);
+    self
+  }
+
+  fn with_errors(mut self, errors: HashSet<TNode>) -> TContext {
+    self.errors = Arc::new(errors);
     self
   }
 
@@ -1038,6 +1062,7 @@ impl Drop for AbortGuard {
 
 #[derive(Clone, Debug, Eq, PartialEq)]
 enum TError {
+  Error,
   Cyclic(Vec<usize>),
   Invalidated,
 }

--- a/src/rust/engine/hashing/Cargo.toml
+++ b/src/rust/engine/hashing/Cargo.toml
@@ -14,7 +14,7 @@ generic-array = "0.14"
 hex = "0.4.3"
 serde = { version = "1.0", features = ["derive"] }
 sha2 = "0.10"
-tokio = { version = "1.21", features = ["fs"] }
+tokio = { version = "1.21", features = ["io-util"] }
 
 [dev-dependencies]
 serde_test = "1.0"

--- a/src/rust/engine/src/scheduler.rs
+++ b/src/rust/engine/src/scheduler.rs
@@ -213,19 +213,20 @@ impl Scheduler {
         .core
         .graph
         .poll(root.into(), last_observed, poll_delay, context)
-        .await?;
+        .await;
       (result, Some(last_observed))
     } else {
-      let result = context.core.graph.create(root.into(), context).await?;
+      let result = context.core.graph.create(root.into(), context).await;
       (result, None)
     };
 
-    Ok((
-      result
-        .try_into()
-        .unwrap_or_else(|e| panic!("A Node implementation was ambiguous: {e:?}")),
+    (
+      result.map(|v| {
+        v.try_into()
+          .unwrap_or_else(|e| panic!("A Node implementation was ambiguous: {e:?}"))
+      }),
       last_observed,
-    ))
+    )
   }
 
   ///
@@ -260,19 +261,13 @@ impl Scheduler {
       results
         .iter()
         .zip(roots.iter())
-        .map(|(result, root)| {
-          let last_observed = result
-            .as_ref()
-            .ok()
-            .and_then(|(_value, last_observed)| *last_observed);
-          (root.clone(), last_observed)
-        })
-        .collect::<Vec<_>>(),
+        .map(|(result, root)| (root.clone(), result.1))
+        .collect(),
     );
 
     results
       .into_iter()
-      .map(|res| res.map(|(value, _last_observed)| value))
+      .map(|(res, _last_observed)| res)
       .collect()
   }
 

--- a/src/rust/engine/src/session.rs
+++ b/src/rust/engine/src/session.rs
@@ -31,7 +31,7 @@ const STRAGGLER_LOGGING_INTERVAL: Duration = Duration::from_secs(30);
 // Root requests are limited to Select nodes, which produce (python) Values.
 pub type Root = Select;
 
-pub type ObservedValueResult = Result<(Value, Option<LastObserved>), Failure>;
+pub type ObservedValueResult = (Result<Value, Failure>, Option<LastObserved>);
 
 ///
 /// An enum for the two cases of `--[no-]dynamic-ui`.

